File: client/python/modeldb/events/MetricEvent.py
"""
Event indicating that a metric has been computed for a model's prediction.
"""
from modeldb.events.Event import Event
from ..thrift.modeldb import ttypes as modeldb_types


class MetricEvent(Event):
    """
    Class for creating and storing MetricEvents
    """

    def __init__(
            self, df, model, labelCol, predictionCol, metricType, metricValue):
        self.df = df
        self.model = model
        self.metric_type = metricType
        self.metric_value = metricValue
        self.label_col = labelCol
        self.prediction_col = predictionCol

    def make_event(self, syncer):
        """
        Constructs a thrift MetricEvent object with appropriate fields.
        """
        syncable_transformer = syncer.convert_model_to_thrift(self.model)
        syncable_dataframe = syncer.convert_df_to_thrift(self.df)
        me = modeldb_types.MetricEvent(
            syncable_dataframe,
            syncable_transformer,
            self.metric_type,
            self.metric_value,
            self.label_col,
            self.prediction_col,
            syncer.experiment_run.id)
        return me

    def associate(self, res, syncer):
        """
        Stores the server response ids into dictionary.
        """
        syncer.store_object(self.df, res.dfId)
        syncer.store_object(self.model, res.modelId)
        syncer.store_object(self, res.eventId)

    def sync(self, syncer):
        """
        Stores MetricEvent on the server.
        """
        me = self.make_event(syncer)
        thrift_client = syncer.client
        res = thrift_client.storeMetricEvent(me)
        self.associate(res, syncer)

File: client/scala/libs/spark.ml/src/main/scala-2.11/edu/mit/csail/db/ml/modeldb/client/event/MetricEvent.scala
package edu.mit.csail.db.ml.modeldb.client.event

import com.twitter.util.Await
import edu.mit.csail.db.ml.modeldb.client.{ModelDbSyncer, SyncableDataFrame, SyncableTransformer}
import modeldb.ModelDBService.FutureIface
import org.apache.spark.ml.Transformer
import org.apache.spark.sql.DataFrame

/**
  * Event indicating that a metric has been computed for a model's prediction.
  *
  * @param dataframe - The DataFrame containing the predictions.
  * @param model - The model that made the predictions.
  * @param labelCol - The column of the true label.
  * @param predictionCol - The column of the prediction label.
  * @param metricType - The choice of metric.
  * @param metricValue - The value of the metric.
  */
case class MetricEvent(dataframe: DataFrame,
                       model: Transformer,
                       labelCol: String,
                       predictionCol: String,
                       metricType: String,
                       metricValue: Float) extends ModelDbEvent {

  /**
    * Store the metric event on the server and do object-ID mappings.
    * @param client - The client that exposes the functions that we
    *               call to store objects in the ModelDB.
    * @param mdbs - The ModelDbSyncer, included so we can update the ID
    *             mappings after syncing.
    */
  override def sync(client: FutureIface, mdbs: Option[ModelDbSyncer]): Unit = {
    val res = Await.result(client.storeMetricEvent(
      modeldb.MetricEvent(
        SyncableDataFrame(dataframe),
        SyncableTransformer(model),
        metricType,
        metricValue,
        labelCol,
        predictionCol,
        experimentRunId = mdbs.get.experimentRun.id
      )
    ))

    // Update object-ID mappings.
    mdbs.get.associateObjectAndId(dataframe, res.dfId)
      .associateObjectAndId(model, res.modelId)
      .associateObjectAndId(this, res.eventId)
  }
}


